### ISSUE TITLE
From embedding Mono to Godot as a library and the future

### ISSUE THREAD

**Issue by neikeq** [godotengine/godot-proposals#2333]
I'm writing this as a GIP as I could not think of a better way to discuss this with everyone. Sorry for not following the GIP template, I don't think it applies well in this case.

Godot supports C# as a scripting language by embedding the Mono runtime. Godot acts as the entry point for the application and takes care of initializing the Mono runtime and loading the required assemblies for execution. This proposal describes an alternative approach.
The idea is for the entry point to be the C# application itself, which then would load Godot as a shared library (via DllImport or [NativeLibrary](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary)). This proposal doesn't pursue the goal of making Godot more usable as a library (which could be great in any case), but only to make it possible to compile Godot as a shared library with an exported entry point for C# to call once at startup.

## Motivation

The main motivation for this change is to reduce the workload needed on Godot's side to support C# on multiple platforms, fix bugs and stay up-to-date with newer .NET versions. By freeing ourselves from that we can focus on more important areas.

Taking only export/publish tooling into account, with this approach the .NET Sdk would already take care for us of Ahead Of Time (AOT) compilation and assembly trimming (of which the latter we don't yet support but would like to). AOT is something we already struggle with and assembly code trimming is a feature we would benefit from almost for free (we will still need to make the Godot C# API and game projects play well with code trimming).

Completely removes the need for us to do the work for building the Mono runtime ([godot-mono-builds](https://github.com/godotengine/godot-mono-builds)) for each target platform and for compiling Godot with it (part of build system code in the mono module, [build-containers](https://github.com/godotengine/build-containers) and [godot-build-scripts](https://github.com/godotengine/godot-build-scripts)). The .NET Sdk already provides the compiled runtimes.

Other benefits:

- Smaller Godot export templates (maybe even none at all?):
  - No longer need to include the Mono runtime and BCL for all target platforms. They are included in the .NET Sdk and their optional workloads.
  - No longer need to include the Godot engine itself as it would be provided as a native shared library via a NuGet package.
- Many existing issues might become obsolete. Issues related to our weird embedding use case where we're likely doing a few things wrong. Issues related to exporting, like native libraries from thirdparty NuGet packages not being included in exported games because Godot doesn't know about them when exporting a game.
- Better IDE support by default, less features need to be implemented in Godot IDE extensions. Projects would be runnable by default as they would be executable rather than library C# projects. A Godot extension would only need to add the more fancy features like `Play in Editor`.
- Benefit from platform-specific APIs in C# (from Xamarin.Android/iOS, etc).
- A less alien experience compared to the rest of the .NET ecosystem. At the very least at the build system level:
  - If you know how to enable AOT, trimming and similar features on C# projects, you know how to do it with Godot too as it's the same.
  - You don't need the Godot editor to export your games. Well, you do, but you can simply run `dotnet publish` and it takes care of everything for you. The plan is to make a MSBuild task that runs the Godot export command to export other Godot assets/PCK (exporting from the editor export dialog would still work too, that doesn't change).
- No more [data folder](https://github.com/godotengine/godot-docs/blob/128d0b4a1574aa835705c7ddce5c6a5df4025984/development/compiling/compiling_with_mono.rst#data-directory) on desktop platforms. .NET can publish self-contained single-file executables for Linux, Windows and macOS. Native libraries are not bundled by default though, so we would have libgodot.so/dll next to the executable. This is not worse than what we have now with the data folder next to the executable. Alternatively we can publish with the property `-p:IncludeNativeLibrariesForSelfExtract=true` to bundle native libraries, resulting now in a true single-file executable. Keep in mind native libraries bundled in the executable are extracted to a temp folder when running the app.

## Downsides

- It's likely games published for the desktop will use the CoreCLR runtime (unless the .NET Sdk provides a way to swap runtimes). This could result in desktop games being bigger. I'm not sure about it though.
  For example, the Mono runtime plus mscorlib, System and System.Core assemblies amount to around 13.1MB without trimming (Linux x64).
  The following is from an attempt to publish a net5.0 hello world app:
  ```
  62.6MB Trim disabled
  21.3MB TrimMode=copyused
  12.6MB TrimMode=link
  ```
  `TrimMode=copyused` is the default trim mode and the safest. It only removes unused assemblies. `TrimMode=link` removes unused members too so it gives the best results ~~but can cause issues with reflection and dynamic code generation~~. _UPDATE: .NET 6 makes link safe to use, plus we may even replace reflection with source generators which makes scripts link safe as well._
  `TrimMode=link` gives us smaller results than we currently have in Godot 3.2, but what we have right now is not trimmed at all so it's not a good comparison. The best would be if the .NET Sdk allowed to choose between Mono and CoreCLR when publishing a desktop app, leaving the choice to the app developer (_EDIT: This might already be possible in .NET 5 but it's not straightforward_).
- Godot provides two ways to export Android games: with a prebuilt apk, or building a custom gradle/java Android project which is fully customizable. If someone is using the latter and after some time they decide to start using C#, then that custom gradle/java Android project needs to be rewritten in C#/Xamarin.Android. Perhaps I'm overestimating the number of developers using this custom Android project or the amount of work needed to move from it, but it would be ideal if a change in scripting language didn't require such work.

## What work needs to be done

Some of the work involved to make this happen:

- Make Godot compilable as a shared library. This library needs at the very least export an entry point that can be called by C#.
- Support CoreCLR. I don't know if the .NET Sdk will allow swaping runtimes. If it's not possible to choose Mono on the desktop, we would need to support CoreCLR in addition to Mono.
  Godot relies on a lot of Mono embedding APIs. Supporting CoreCLR would require finding alternatives and/or moving many of those parts to C#. For C# to C++ calls we would use P/Invoke instead of Mono internal calls. P/Invoke doesn't allow to pass managed objects (we can pass gchandles though if needed). This may force us to rethink a few internals. _UPDATE: We will be moving most interfacing code to C#. We will eventually use function pointers (C# 9 feature) instead of P/Invoke. Godot will not use any embedding APIs so it will work with all runtimes: Mono, CoreCLR and the experimental Native/CoreRT._
  ~~_NOTE: I think supporting CoreCLR would be needed any way if we want to support UWP. AFAIK Mono doesn't support that? But don't quote me on that!_~~ Never mind, Mono can run on UWP as commented below.
- For C# project we need to override the export process to work with MSBuild/`dotnet publish`.
- Distribute NuGet packages for the Godot C# API and the Godot native libraries.

## When would this work begin

As soon as it's approved if that happens. As for the Godot version milestone, it depends whether Godot 4.0 is released before or after a .NET 6 release candidate. Preferably we want to use .NET 6, as .NET 5 doesn't support mobile. But if that isn't possible we have 3 options:
1. Implement this proposal with .NET 5 in Godot 4.0, with mobile support for C# delayed to Godot 4.1.
2. None of the above. Stay as we are for Godot 4.0, embedding Mono with the classic Mono BCL instead of .NET 5/6 support.
3. Release Godot 4.0 without C# support (either skip 4.0+C# or delay its release). I honestly prefer this over option 2 as I don't want to introduce a new major version that's known to still be going to cause trouble and needs to be supported for some time (we already have 3.2.x which has many pending fixes). And it doesn't feel right to then introduce a minor version that breaks both backward and forward compatibility... We've been doing this with C# in 3.x under the alpha/beta state excuse, but I wanted to stop that with 4.0.

_**UPDATE:** The final decision is that Godot 4.0 will stay as is and this proposal will be implemented in Godot 4.1._

We're approaching the right time for a change like this. The work involved in switching to .NET 6 just to get back to our current state would likely be the same work if not more than the implementation of this proposal involves. That's without adding support for more platforms and features, which currently takes a lot of time and this proposal aims to simplify considerably.

**Comment by NHodgesVFX** [comment]
This looks amazing super excited for coreclr. Like I mentioned in chat I think it would be a good idea to make use of system.numurics wherever possible for the speed benefit. Maybe we can also make use of spans and the hardware accelerated math(in .net 6, we might get this for free if we use the relevant methods already). Although this could be added later if this part of the code doesnt need to be reworked for this proposal.  

For the last part I believe Microsoft is targeting September for the release candidate. I don't think Godot would be released before then but as a fallback I vote for 1 as it doesn't leave c# users with no c# and it should be relatively easy to bump to .net 6 when its stable.

**Comment by Xartorx** [comment]
Looks awesome, ways better than current system. 
About Android - there's JNI, but using it comes with its can of worms.

>Make Godot compilable as a shared library. This library needs at the very least export an entry point that can be called by C#

Could this be also good for other languages, ex. Rust or D?  Because currently using GDNative is cumbersome, and using Godot as library would be easier than setting up GDNative and messing with `gdns` and `gdnlib` files.

**Comment by Calinou** [comment]
> Could this be also good for other languages, ex. Rust or D? Because currently using GDNative is cumbersome, and using Godot as library would be easier than setting up GDNative and messing with `gdns` and `gdnlib` files.

GDNative is being rewritten for 4.0. It should be more convenient already: https://github.com/godotengine/godot/pull/44989

**Comment by Xrayez** [comment]
> I'm writing this as a GIP as I could not think of a better way to discuss this with everyone. Sorry for not following the GIP template, I don't think it applies well in this case.

We must move to GitHub Discussions for this kind of workflow, see my meta proposal at #2069.

**Comment by migueldeicaza** [comment]
This proposal has a bunch of different elements to it, and I think it should be broken up.

1. Consuming Godot as a shared library
2. Build changes for different form factors
3. Adding an option for using coreclr

Some observations:

# On a Shared Library

This to me, it's probably the most interesting part of this proposal. There are several advantages of exposing Godot as a shared library, in addition to the points stated above, it can be used to easily embed the engine into existing applications that may have been written in other languages. For example, it could be used to augment an existing mobile application, or argument that's the publication with some content that was produced inside Godot.

Providing the engine as a shared library would enable scenarios we are game content can be as easily added to an application in the same way that webpages added to applications.

It is fun to me that this is being discussed now, many years ago I reached out to Juan inquiring precisely about this capability. At the time we wanted to allow developers to a 3-D content to their applications across multiple platforms (We were trying to provide across the platform 3-D engine that users could use).  At the time this was not possible, so we end up picking up another open-source engine, it worked great, and many developers were able to create xplat solutions this way.  The engine that we chose at the time is no match for the capabilities of Godot, and sadly no longer being maintained, so I am personally interested in getting a replacement for users.

Today there is a big gap in our industry for a cross-platform 3-D solution. I believe that this would solve problems for our large segment of developers.

# On VM work

* The API for .NET 6 is breaking compared to the mobile profile in .NET
* New core APIs like span and numerics will be available on both runtimes (mono and coreclr)
* Mono can be used in UWP
* Easiest way to use coreclr would be to clone the embedding API of mono for coreclr (we did a proof of concept once), this ensures you still work on wasm, mobile and consoles which only work in mono
* if you are concerned about perf then you should try using the LVN backend for Mono 

I would choose option (2) at least for a while, as there will be changes in the embedding APIs in this run times in the future so you might want to wait until those are sorted out.  Additionally, the word that you need to swap run times and much of the work that is happening in .net 6 is still under development and will take a few more months before we even know what the shape of it is. I am personally a fan of limiting the amount of moving pieces, 4.0 it's a major change already.

**Comment by aaronfranke** [comment]
I don't like option 3 because it would make upgrading to 4.0 not an option for many people and would probably cause community outrage. Option 1 is not really ideal either. I think the best choice is option 2.

As Miguel says, there are still changes that need to be sorted out, and it doesn't make sense to me to switch to .NET 5/6/etc until those are sorted out (or otherwise we might have to "switch to it twice"). Also, people are going to have a hard time converting their projects from 3.2 to 4.0 anyway, so having 4.0 use Mono can be a blessing in disguise, and then C# users can just do additional work upgrading their projects to 4.1/4.2/etc.

It's OK if compatibility breaks with C# in a minor release in 4.x, we don't claim that the C# API is stable yet, and we have already broken C# compatibility in *patch releases* several times in 3.0.x and 3.2.x, so that would still be an improvement :)

**Comment by fab918** [comment]
For the shared library idea, I think mostly everyone agrees.

For the .net version:

As:
- Godot 4.0 should not arrive before the end of the year (They told in the last Q&A the alpha for godot 4 should come in April at best)
- .NET 6 preview is already out, and the release in planned to November
- .NET 6 is the replacement for .net core/.net framework
- .NET 6 is an LTS version

I think the best move is to focus on .NET 6 instead of investing time on a technology that is officially “dead” and loose time. Plus, the chance are very high that .Net 6 will be out before the Godot 4.0 release, and even in the worst case, I think the c# community will prefer to wait 1-2 months more to get .NET 6, it's a better investment for the future.

For CoreCLR runtime, it's supposed to replace at the end mono runtime, it seems a good to invest on it, plus the size difference is low (event if being able to choice is always a plus). @neikeq , However .NET 6 not supposed supported both runtime out of the box (or I missed something)?

![image](https://user-images.githubusercontent.com/56844494/109420964-99c8ca00-79d5-11eb-8077-5e7b32a46ad5.png)

**Comment by decepticlown** [comment]
If this is implemented in future, I'm talking about Godot as library. Will it open possibility for adding c++ scripts more easily? Like without GDNative?

**Comment by migueldeicaza** [comment]
Hello,

Mono is the VM that supports Android, iOS, consoles and WebAssembly, and also part of the .net 6 platform.  Switching to CoreCLR as the VM means none of those platforms will be supported by Godot.

**Comment by Delpire** [comment]
@migueldeicaza - What exactly does that mean? I'm still not very clear on what .NET 6 is supporting. It seems like more and more of the Mono runtime is going into the "core" .NET. .NET 5 only has support for Windows, MacOS, and Linux, but it looks like .NET 6 is adding support for Android, iOS, Apple's silicon, and Windows Arm64.

My understanding was that in addition to the netx.x targets (i.e. net5.0, net6.0), there will be a net6.0-android and net6.0-ios target. Part of the vision of MAUI was that you could have a single application that targeted multiple OS's (except Linux 😅). In Godot you have to export specific targets, would targeting those TFM's not be enough in .NET 6?

The only thing I see as missing is console and WASM support, thought I think WASM support must exist in .NET Core and .NET 5/6 since we have client side Blazor. (I believe its like .NET Core assembles being interpreted by a Mono WASM runtime or something like that).

Whether it be Mono or CoreCLR, I think the biggest gap in Godot's C# support (at least for me), is being able to use all the .NET tooling I use on a daily basis anyway. Being able to run my game from Visual Studio, debug without having a third party extension, use Visual Studio's profilers, or tools like PerfView. Maybe things would be less bad if Mono profiling for Godot wasn't broken on Windows, and maybe the problem is Godot not existing as a shared library, I'm not entirely sure, but I think whatever direction Godot goes in, y'all should think about those use cases. I already feel really efficient using C# with Godot, but if using Visual Studio and Godot at the same time was as seamless as working on a Blazor application, or a .NET 5 console app, it would be a big win.

**Comment by NHodgesVFX** [comment]
> Hello,
> 
> Mono is the VM that supports Android, iOS, consoles and WebAssembly, and also part of the .net 6 platform. Switching to CoreCLR as the VM means none of those platforms will be supported by Godot.

This is incorrect. CoreCLR will support those platforms by using mono in some form. Basically as @Delpire said. On consoles its should run just fine as most are based on linux look at unity which uses c# and runs on consoles just fine.

**Comment by neikeq** [comment]
Reiterating what the previous replies have said, .NET 6 will have 2 runtimes: CoreCLR and Mono. Mono is the runtime that powers .NET 6 on mobile and wasm. The plan is for Godot to support CoreCLR in addition to Mono, not as a replacement.

**Comment by aaronfranke** [comment]
Long-term, does Microsoft have plans to make CoreCLR run on mobile and WebAssembly? Having different runtimes for different platforms doesn't seem like a good long-term solution (for either Microsoft or Godot).

**Comment by Delpire** [comment]
@aaronfranke - After some more reading, my understanding is that (as of 2019) Microsoft wants to keep Mono and CoreCLR as separate runtimes, with Mono focused on iOS and Android, and CoreCLR focused on desktop. They want to make the two of them drop in replacements for one another that you can control with a build switch. I think if this is what they envision for .NET, Godot should adopt it, and it should be possible to tell Godot which target to build. We already have separate exports for Windows, MacOS, Linux, Android, etc., they should just be able to change the target from net6.0 to net6.0-iOS or net6.0-android.

Its possible there will be APIs that still require either Mono or CoreCLR and not work on the other, if swapping between targeting net6.0 and net6.0-iOS is as easy as a build switch, then Godot users can just use conditional compilation for those iOS specific calls.

**Comment by neikeq** [comment]
>Having different runtimes for different platforms doesn't seem like a good long-term solution (for either Microsoft or Godot).

For Godot, other than the initial effort to support the new runtime, it should not be an issue in the long-term. The idea is to move more code from C++ to C# and avoid dealing with each runtime API as much as possible. At some point we may also want to support [CoreRT](https://github.com/dotnet/runtimelab/tree/feature/NativeAOT).

**Comment by GeorgeS2019** [comment]
> Easiest way to use coreclr would be to clone the embedding API of mono for coreclr 

@migueldeicaza @neikeq  based on the [recent test of coreclr dotnet cli ( NET5.0.5)](https://github.com/godotengine/godot-proposals/issues/793#issuecomment-814989441) on a godot project that run flawlessly with the non dotnet cli godot build => I have reservation how easy later to clone the embedding API of mono to CoreClr => without iteratively feedback to the coreclr team on what else are still missing in .NET5/6 needed by desktop software (e.g. godot for desktop) based on mono.

**Comment by neikeq** [comment]
@GeorgeS2019 We won't be using the embedding APIs anymore as we'll be moving the code that relies on them to C#.
The Godot editor will still use embedding, but it will only use the CoreCLR host API to start the runtime, and load our helper assembly. I had a look at the Mono NETCore App runtime packages and it seems to export the CoreCLR host API, so the same code may even work for both runtimes.

**Comment by nobuyukinyuu** [comment]
If these changes make it easier to profile code written in c# I'm all for it.  It seems extremely non-trivial to profile code with the current mono build in a way that's as easily accessible as either the Godot profiler or the VS one...

**Comment by GeorgeS2019** [comment]
@nobuyukinyuu   
- It seems, instead of the current 3 steps to get to a Mono Godot4 build, it means ONE step with this new merge
- It is unclear how to track the bugs that come from the 67 merged file changes without mono unit tests and sufficient integration test code coverage provided by the existing godot 4 demo in c#
- Do we need a clearer discussion what comes after First step, what are second and third?

**Comment by GeorgeS2019** [comment]
>[from @migueldeicaza]  many years ago I reached out to Juan inquiring precisely about this capability. At the time we wanted to allow developers to a 3-D content to their applications across multiple platforms (We were trying to provide across the platform 3-D engine that users could use). At the time this was not possible, so we end up picking up another open-source engine, it worked great, and many developers were able to create xplat solutions this way. **The engine that we chose at the time is** no match for the capabilities of Godot, and **sadly no longer being maintained**, so I am personally interested in getting a replacement for users.

The gossip among .NET developers following the engine referred my Miguel is that .NET c# 3D engine with MIT open source is a dream that is full of many challenges that are often difficult to comprehend.

We now have a chance with Godot c# 3D engine. 

I **urge** developers who have gone through decades SEEING failures (one project after another .NET project) involving truly open source (MIT) 3D .NET c# engine to COME together to make this [Godot .NET6) WORKS! Let us plan the path to .NET6 with robust Unit and integration tests and code profiling!

In the last Godot Q&A (youtube), there were many questions on using Visual Studio e.g. 2019 as IDE for unit/integration tests and code profiling.  If this issue is important to you, do join on regular basis to feedback to the Godot Team.

**Comment by aaronfranke** [comment]
@GeorgeS2019 Please be patient. Regardless of your urges and many posts, these things will take time.

**Comment by realvictorprm** [comment]
As much as I like the love for C# here I would like to add that it would be very good if the implementation doesn't limit the support to one specific language running on the CLR but works for all languages which can run on the CLR.  My language of choice would be F# which doesn't work that well with Unity as it could work therefor I have high hopes for this here :)

**Comment by realvictorprm** [comment]
Side note, I don't ask for specific F# support with own api etc, just asking to do things such that the used laguage is not limited to be C# (which the other language can compensate via interop to a C# API)

**Comment by neikeq** [comment]
> As much as I like the love for C# here I would like to add that it would be very good if the implementation doesn't limit the support to one specific language running on the CLR but works for all languages which can run on the CLR. My language of choice would be F# which doesn't work that well with Unity as it could work therefor I have high hopes for this here :)

I do want to make the implementation language agnostics to allow for third-parties to add other CLR languages with the same features as we have for C# and without workarounds like having an empty `.cs` file with the same name as placeholder. However first we need to focus on good C# support.

**Comment by voronoipotato** [comment]
So long as it is a library we F# folks can write a shim, we are somewhat used to higher friction. Right now as a near term solution I'm trialing leveraging GDNative with F# to avoid these kinds of workarounds and being constrained to just using mono. So this topic is pretty exciting to me.

**Comment by realvictorprm** [comment]
So I would like to way in here and help out getting this stuff done sooner :) however I'm not sure how much effort it requires and whether it makes sense for me as my time unfortunately is quite limited because now that I work full time I have way less time :/ let me know whether you're happy to have additional resources :)

**Comment by GeorgeS2019** [comment]
> Many **(.NET6) features** are in close-to-final form and others will come soon now that the **(.NET6) foundational building blocks** are in place for the release. [.NET6 Preview 4](https://devblogs.microsoft.com/dotnet/announcing-net-6-preview-4/) establishes a solid base for delivering a final .NET 6 build in November, with finished features and experiences.

Is now the time for us to plan a path for Godot4 to .NET6 path based on the fact that "ALL! the foundational building blocks of .NET6 are now in place?

**Comment by reduz** [comment]
Just a follow up from my side. With @vnen we are rewriting GDNative entirely and converting it into a low level extension API (no longer a script API). May also probably rename it to GDExtension avoid confusion. 

With the new API you will be able to do pretty much everything you do with modules, such as registering objects to class DB, and the entire Godot API is a single C function pointer away. We still have some work to do, but I think it may be interesting to evaluate if .NET makes more sense to use this, as it has the potential to simplify the code a lot to make it hot-pluggable.

The general idea is to leave the script API to dynamically typed languages such as GDScript or VisualScript, while compiled languages work via the extension system. This way, the GDnative C++ and other bindings such as Rust, Nim, etc. will no longer work like scripts, and instead like modules.  This means you register a class and thats it, it appears in the engine ready to use as a node, resource, etc. with documentation and everything. 

The main rationale behind this is that .NET is compiled, for which it makes more sense it acts like a module/extension than a scripting language.

**Comment by Zarmot** [comment]
> The general idea is to leave the script API to dynamically typed languages such as GDScript or VisualScript, while compiled languages work via the extension system. This way, the GDnative C++ and other bindings such as Rust, Nim, etc. will no longer work like scripts, and instead like modules. This means you register a class and thats it, it appears in the engine ready to use as a node, resource, etc. with documentation and everything.

This may also be helpful for cross language interoperability, "Register a class" can make it more convenient to use it in GDScript. And that reducing the "black box" in cross language interop witch can help developers better understand what happened. For example, I like to use C# to write core logic as a library, and use GDScript to build other parts(GUI, simple logic, etc), I love the dynamics of GDScript, the above changes can make the whole process more controllable.

Considering that one of the meanings of supporting C# is to attract developers from unity, maybe it's helpful to encapsulate a script API based on this？

**Comment by neikeq** [comment]
While the new ClassDB extensibility API sounds great, it's not really encompassed in this proposal so let's not discuss it here.
If any one is wondering though, I do want to support that in C#.

**Comment by GeorgeS2019** [comment]
There is a BIG difference between .NET6 ecosystem and the topic discussed here ( which is only a subset of .NET6 ecosystem discussion)

![image](https://user-images.githubusercontent.com/49812372/121595355-7b478580-ca3e-11eb-818a-37310c9f9b00.png)

**Comment by Shadowblitz16** [comment]
> This means you register a class and thats it, it appears in the engine ready to use as a node, resource, etc. with documentation and everything.

@reduz are you saying we will finally be able to make true custom node types?
if so I'm in.

**Comment by aaronfranke** [comment]
@Shadowblitz16 Yes, that's the idea. There is also discussion of whether we should make that the only option, so every script created in GDNative (and maybe every script created in C#) shows up and behaves just like an actual built-in engine type.

**Comment by realvictorprm** [comment]
Would you folks be open for contributions which as soon as this arrives adds basic F# support in the editor?

**Comment by Calinou** [comment]
> Would you folks be open for contributions which as soon as this arrives adds basic F# support in the editor?

The Godot script editor doesn't have full C# support in the first place, so F# support isn't a goal right now. Adding full C# support to the Godot script editor would require turning it into a LSP *client*, which is a lot of work.

**Comment by aaronfranke** [comment]
@Calinou I'm pretty sure @realvictorprm was just talking about things like allowing creating F# scripts from the create script dialog and allowing Godot to compile and run them.

**Comment by realvictorprm** [comment]
Exactly @aaronfranke, this would be the first step.

Besides, as fancy as it sounds to have full F# scripting support in the engine, as bad is the idea to actually implement it in my opinion. I think the engine should not focus on having a geat editor as we already have great editors out there  (VSCode, IntelliJ, Atom etc.). More importantly the engine's build system should be in such a state, that it's possible to easily integrate build scripts for additional languages  + easy ways to seamlessly integrate the workflow of developing in an external editor or IDE of your choice.
Therefor I would focus my effort on the build system side of the engine (e.g. adding the dialog to add F# files) and in that process also trying to ensure that it's working with multiple languages (if it's a feature which should be supported language independ).

The only thing which I would support/try additionally, as it's not too difficult (famous last words 😄), is simple syntax highlighting for F# scripts, but that's it. For anything more you would need support for the LSP, which, as already outlined by @Calinou, is a huge effort.

**Comment by chismar** [comment]
So, can anyone provide some update on how the process is going on? What's done, what's to be done? I know that it can't exactly be finished before .net 6 releases, but at least to understand how it progresses would be nice.

**Comment by neikeq** [comment]
The decision was to postpone this proposal (including .NET 6 support) until Godot 4.1. Not much work has been done yet other than starting to move code to C#. All the work in progress will go into the [dotnet6](https://github.com/godotengine/godot/tree/dotnet6) branch.

As for what's to be done, the only steps that are clear so far are:

- Move code to C#.
  - Marshaling and internal calls. Mostly done with godotengine/godot#48409, pending some code.
  - Rest of the code. Several parts from `csharp_script.cpp` and `gd_mono*.cpp` files.
  - At this point I may start playing more with source generators. For example, I want to replace reflection + marshaling type checks for invoking methods and accessing properties on scripts from the engine.
- Replace embedding Mono with CoreCLR hosting/embedding and introduce .NET 6 support. Only for the editor, as export templates will be used as libraries instead of embedding as this proposal explains. This doesn't mean we stop using Mono as explained in previous comments.
  - Export templates will stop working for now.
  - Assembly reloading will stop working for now (AppDomains no longer supported).
- Re-add assembly reloading with AssemblyLoadContext.

Everything else will be decided as things progress.

**Comment by viksl** [comment]
I'm just curios but was there or could there be a contact with miscrosoft about a possible grant with this project in mind since they are pushing the latest and neweswt net6?
Well, I'm not sure if this question belongs here so I apologize if it's off-topic.

**Comment by reduz** [comment]
Will clarify a few things:
@neikeq The idea in Godot 4.0 onwards is that compiled languages can extend the engine, and extensions can be loaded and unloaded at any time, when obtained from asset library.

This opens the door at users writing extensions for other users in any language they like (C++, Rust, Nim) then you get shared objects that you load into the engine and immediately extend the engine API and provide native nodes, not scripts.

Making Mono work by controlling the Godot entry point fundamentally goes against this vision, as it will make the language not play along the others nicely. If most users preferred C# and this was the unique language available, it would make sense but currently it sits around 20-30% of the userbase, the rest strongly prefers to use GDScript, C++ and a smaller niche prefers other languages such as Rust or Nim.

I am not against the fundamemtal idea of the proposal, as it can definitely be of use to some, but it wold strongly exclude others if it this was the default way of using it.

**Comment by neikeq** [comment]
> The idea in Godot 4.0 onwards is that compiled languages can extend the engine, and extensions can be loaded and unloaded at any time, when obtained from asset library.

The new extensions API is very great for many reasons and I plan to add support for it in C# (albeit I still wish it allowed namespaces). However, those benefits are mainly for native language bindings that used GDNative. The C# bindings already supports all that.

> In other words, the idea is that, if some user makes a fantastic terrain plugin in C#, and wants to make it available to others using GDScript or C++ for their game, the plugin should install normally and allow other users to continue using Godot as always.

When it comes to the Godot editor this GIP doesn't change much as the editor will still use embedding as it does now. Godot would be used as a shared library only in exported games. Regarding those, when I wrote this I kept in mind that nothing should break for projects using a different language that at some point decide to start using C#. Right now the only caveat is Android (and perhaps iOS), which I mention in the downsides:

> - Godot provides two ways to export Android games: with a prebuilt apk, or building a custom gradle/java Android project which is fully customizable. If someone is using the latter and after some time they decide to start using C#, then that custom gradle/java Android project needs to be rewritten in C#/Xamarin.Android. Perhaps I'm overestimating the number of developers using this custom Android project or the amount of work needed to move from it, but it would be ideal if a change in scripting language didn't require such work.

But who knows, maybe with some extra effort I may be able to keep the prebuilt apk and the java gradle project working with C#.

> Right now this is not possible because using C#/Mono requires a separate engine build, so I would put most of my effort to make sure that this is made possible via the new extension system, rather than focusing on the approach suggested in this proposal.

Although not formalized as a GIP, there's a plan to include C# support without needing a separate Godot editor build. This GIP doesn't prevent that. It even moves us a bit in that direction by not bundling a .NET runtime and its framework libraries with the Godot editor.

**Comment by reduz** [comment]
@neikeq My point is mostly that you will have users wanting to use an add-on made with C#, and they don't care about what's going on with C# or launching Godot as a library. If this helps somehow while making an actual project in C#, it can eventually be added, but the priority should be to make it work like the rest of the scripting languages. Additionally, if we support this in C#, it is possible other languages will want to go this route, making the system incompatible between them.

The main advantage of the extension system is that it pretty much allows you to use Mono in stock Godot and stock export templates, which is not possible today, It is not a replacement of the script API, but a lower level way to interact with the engine.

Regarding namespaces, I see no problem implementing some support for this, but from the extension API it sounds like it will most likely just be no more than adding a namespace string while registering the extension class.

**Comment by andreakarasho** [comment]
just FYI:
using [NativeAOT](https://github.com/dotnet/runtimelab/tree/feature/NativeAOT) you can static link libraries in the csproj:

```xml
<ItemGroup>
 <NativeLibrary Include="SDL2.lib" />
</ItemGroup>
```

**Comment by GeorgeS2019** [comment]
just FYI:
**[3D engine as an embedding library](https://github.com/microsoft/microsoft-ui-xaml/issues/5156#issue-913822617) for** the project reunion of **.NET6**

**WinForm for Visual Basic and c#** remains one of the biggest developer segments for Window. 

Now finally an embedding 3D engine library is available as part of the **Project reunion of .NET6.** This opens out 3D game/visualization for **WinForm/WPF/UWP developers** with **modern Windows technologies and new features, plus the best of existing Desktop (aka Win32) features**

**Comment by tapir** [comment]
There hasn't been much updates in this proposal, is it still the case that this will come out with 4.1 and Godot as a .NET 6.0 library?

**Comment by daerogami** [comment]
To add to @tapir's follow up question. @reduz mentioned a GDNative rewrite would provide possible support as a shared library. What issues/PRs is that effort captured by?

**Comment by GeorgeS2019** [comment]
@tapir 

>This [PR ](https://github.com/godotengine/godot/pull/56303#issue-1090163973 )introduces many changes before the upgrade to .NET 6 (from the current .NET 5). After this PR is merged I'll rebase the dotnet6 branch and then push a few other important changes before the **dotnet6 branch can be merged into master.**

@daerogami  => sounds like thing will soon speed up.

![image](https://user-images.githubusercontent.com/49812372/150205949-18256771-1b6e-4021-9498-1a09bc74dd55.png)

**Comment by GeorgeS2019** [comment]
@NHodgesVFX 

Tentatively, I imagine we need a[ JavaScript -Godot Bridge](https://www.youtube.com/watch?v=Q8G3oyZwbsE&t)

This will eventually involve Blazor JavaScript interop 

Sharing here so we can gather discussion.

However, it is best to create a separate proposal once there is sufficient interest.

**Comment by GeorgeS2019** [comment]
[First attempt to merge .dotnet6 branch to master has started.](https://github.com/godotengine/godot-proposals/discussions/4007#discussioncomment-2265117)

**Comment by GeorgeS2019** [comment]
@andreakarasho

Godot 4 .NET6 C#: Initial [**NativeAOT**](https://github.com/godotengine/godot/commit/a46e601d491ce3105312a0f93b504876d55cc798) support

**Comment by GeorgeS2019** [comment]
![image](https://user-images.githubusercontent.com/49812372/156886886-4a503df0-47c1-4279-a3d3-74e1a0585e0f.png)

[First attempt to merge .dotnet6 branch to master has started.](https://github.com/godotengine/godot/compare/58970df526ad...8b85bc71595e)

**Comment by GeorgeS2019** [comment]
A few of us are attempting to make Godot4 .NET6 (dotnet6 branch) running and we are getting good support from the Godot Team!

https://gist.github.com/lewiji/2b86f47dcb142a10e104504a343c38ff

**Comment by GeorgeS2019** [comment]
Below is a feedback from a few of us who recently attempting to understand how the on-going Godot4 dotnet branch works. This is a draft and covers only a small part of the proposed alternative approach. We hope more users will come forwards and take what we have done to [different topics for discussion](https://github.com/godotengine/godot-proposals/discussions) to speed up the process (community coordinated) needed to support the Godot team to complete the process of migration from embedding mono to .NET6 runtime.

The preliminary findings with feedback from the Godot team and other effort
- [Godot dotnet6 branch notes](https://gist.github.com/lewiji/2b86f47dcb142a10e104504a343c38ff)
- [Dotnet interop playground: ](https://github.com/glihm/dotnet-interop-pgnd) attempt to understand how NativeAOT works in Godot4 dotnet6 context

> Godot **supports C#** as a scripting language **by embedding the Mono runtime**. 
> - **Godot acts as the entry point** for the application and 
> - Godot takes care of initializing the Mono runtime and loading the required assemblies for execution. 

>This proposal describes an alternative approach.

> - The alternative involves c# application as the entry point and no longer the Godot
> - The C# application would load Godot as a shared library (via DllImport or [NativeLibrary](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary)). 
> - The C# application takes care of initializing the Godot shared library (`'__Internal'`) through the exported entry point for c# to call once at startup
> - The Godot shared library is now "embedded" as part of GodotSharp.dll and will be provided as Nuget.
> - [dotnet6 commit : Make GodotSharp API a NuGet package](https://github.com/godotengine/godot/commit/8b85bc71595e1b95f66857d13a6558d492505927)

<img src="https://user-images.githubusercontent.com/49812372/159661888-4e00094d-7a75-43a6-8fed-bdfaa89a933a.png" width="800" />

<img src="https://user-images.githubusercontent.com/49812372/159660795-5694749f-6172-4cee-a415-845fa37a5974.png" width="400" />

> ## Motivation

> The main motivation for this change is to reduce the workload needed on Godot's side to support C# on multiple platforms, fix bugs and stay up-to-date with newer .NET versions. By freeing ourselves from that we can focus on more important areas.

> With this alternative approach, the Godot team no longer takes care on maintaining Mono issues associated with supporting multiple platforms: building the Mono runtime ([godot-mono-builds](https://github.com/godotengine/godot-mono-builds)) for each target platform and for compiling Godot with it (part of build system code in the mono module, [build-containers](https://github.com/godotengine/build-containers) and [godot-build-scripts](https://github.com/godotengine/godot-build-scripts)). 

> The alternative approach means the .NET Sdk provides the compiled runtimes for multiple platforms.

#### (Some of the listed) other benefits:
> - Smaller Godot export templates (maybe even none at all?)
> - Better IDE support by default, less features need to be implemented in Godot IDE extensions. Projects would be runnable by default as they would be executable rather than library C# projects. A Godot extension would only need to add the more fancy features like Play in Editor.
> - Benefit from platform-specific APIs in C# (from Xamarin.Android/iOS, etc).

#### How Godot .NET community could START contribute ? 
Bring the following points to [discussion](https://github.com/godotengine/godot-proposals/discussions)?

- Follow the godot team's [C# Godot 4.0 Roadmap](https://github.com/orgs/godotengine/projects/30/views/1)
- Follow and make coordinated user initiatives: [e.g. Godot dotnet6 branch notes](https://gist.github.com/lewiji/2b86f47dcb142a10e104504a343c38ff) on what PoCs have been done and shared.
- Currently the available dotnet6 branch commits focus on desktop. Support and coordinated initiatives are needed to other platforms (e.g Android/iOS etc) and PERHAPS the feasibility for supporting MAUI or Blazor. 
- Extend one user [initiative ](https://github.com/glihm/dotnet-interop-pgnd) to understand how the [Godot dotnet6 NativeAOT ](https://github.com/godotengine/godot/commit/a46e601d491ce3105312a0f93b504876d55cc798) works for desktop to mobile or perhaps web 
- Share ideas for better IDE experience for development and better Godot dotnet6 specific testing framework made possible by the alternative .NET6 approach

**Comment by GeorgeS2019** [comment]
For many waiting for the next batch of dotnet6 commits scheduled for March/April 2022, here is a preliminary list

![image](https://user-images.githubusercontent.com/49812372/161202740-d07f1856-a43b-40ea-a120-3b25f9ec52bf.png)

**Comment by kopkako** [comment]
Not sure if this is the right place to post this, but I'd like to add:
If there is any way to add the new `dotnet watch` feature that enables **Hot Code Reload**, allowing for changing the program while it is running – that would be *amazing* for Godot development and make me and probably everyone using C# with Godot very happy.

Immediate feedback is a killer feature, really great for creative work and something I'm missing when using C# over GDScript, so yeah, would be great to at least see if this could be possible.

**Comment by kopkako** [comment]
Sorry for the double post, but I think this warrant a separate post:

Could someone clarify what the current plan of action is regarding the future of C# and Godot 4.X?
I read through the entire issue and am still confused, seems to me like there has not been a decision made yet?

- Will .NET 6 be implemented in the same fashion as the Mono Runtime was implemented until now?
- Will this use the new GDExtension API, or will it continue to be a separate Editor binary like it was/is with Godot Mono?
- Is the plan to implement this for 4.0 or 4.1? (In my opinion waiting for 4.1 would be really disappointing for the substantial amount of devs currently using C#, ~20% is a lot!)

Would love to keep on using Godot with C# in the future and really just trying to understand where we're headed.
Maybe @neikeq can clarify?

**Comment by GeorgeS2019** [comment]
@kopkako 

I only speak as someone outside the Godot team who believes in growing the Godot .NET community.

> I will not feedback on the questions you raised **directly** but using this opportunity to lobby for **MORE** Godot .NET enthusiasts to step forwards like yourself **to voice your opinions.** 

> As a Godot .NET user, I could not yet now understand with sufficient overview the challenges of adopting the GDExtension API and what it means in terms of **how long further we need to wait for** BEFORE we have a working and up-to-date with the master branch environment like GDScript 2.0 but for Godot4 for .NET developers. 

> I see the following options and which directions we will go depending **a lot on more feedback** from the Godot .NET community to step forwards to share their views.

> A few weeks back (~15th March 2022), a few of us get our feet wet in learning more about the status of the Godot4 dotnet6 branch because we have similar concerns as @kopkako 

**Option 1**:  In an ideal case, Godot4 support both the vision of GDExtension API and the existing .NET6 path side by side in ways that are synergistic: both paths co-exist in a win-win way. 

> First, we can start discussing the feasibility of option 1 NOW without waiting. 

> Although the existing dotnet6 branch is not up-to-date as the master branch, it has provided the possibilities to **start thinking/discussing, testing (Linux and Windows), and even possibly planning** WHAT COULD BE the benefits of staying and keeping the .NET6 path, **in co-existent** with the GDExtension. This requires more Godot .NET users to step forwards.  For example, how likely to implement `dotnet watch`? If so, how long do we need to wait to get there?

**Option 2**: eventually there will be ONLY GDExtension for c# developer.

> At some point, after the merging of dotnet6 branch, it will transition FULLY into the GDExtension API. 

> It is unclear how feasible is that and how easy to keep the benefits of having a c# application as the entry point which opens to the benefits (partly or fully) of the .NET6/NET7 ecosystem. 

I encourage more .NET users to step forward to share your view!

**Comment by kopkako** [comment]
> @kopkako
> 
> I only speak as someone outside the Godot team who believes in growing the Godot .NET community.
> 
> > I will not feedback on the questions you raised **directly** but using this opportunity to lobby for **MORE** Godot .NET enthusiasts to step forwards like yourself **to voice your opinions.**
> 
> > As a Godot .NET user, I could not yet now understand with sufficient overview the challenges of adopting the GDExtension API and what it means in terms of **how long further we need to wait for** BEFORE we have a working and up-to-date with the master branch environment like GDScript 2.0 but for Godot4 for .NET developers.
> 
> > I see the following options and which directions we will go depending **a lot on more feedback** from the Godot .NET community to step forwards to share their views.
> 
> > A few weeks back (~15th March 2022), a few of us get our feet wet in learning more about the status of the Godot4 dotnet6 branch because we have similar concerns as @kopkako
> 
> **Option 1**: In an ideal case, Godot4 support both the vision of GDExtension API and the existing .NET6 path side by side in ways that are synergistic: both paths co-exist in a win-win way.
> 
> > First, we can start discussing the feasibility of option 1 NOW without waiting.
> 
> > Although the existing dotnet6 branch is not up-to-date as the master branch, it has provided the possibilities to **start thinking/discussing, testing (Linux and Windows), and even possibly planning** WHAT COULD BE the benefits of staying and keeping the .NET6 path, **in co-existent** with the GDExtension. This requires more Godot .NET users to step forwards.  For example, how likely to implement `dotnet watch`? If so, how long do we need to wait to get there?
> 
> **Option 2**: eventually there will be ONLY GDExtension for c# developer.
> 
> > At some point, after the merging of dotnet6 branch, it will transition FULLY into the GDExtension API.
> 
> > It is unclear how feasible is that and how easy to keep the benefits of having a c# application as the entry point which opens to the benefits (partly or fully) of the .NET6/NET7 ecosystem.
> 
> I encourage more .NET users to step forward to share your view!

Thanks for your reply!

I share many of those sentiments and would like to add:
I believe C# was and still is the ideal candidate for a compiled, type-safe & fast language besides GDScript, for more advanced users. I would personally not want to build complex games with GDScript, coming from Unity – having C# as a first-class option is really important to me, so I would like to see it integrated as well as possible.

Of course, the point of contention is: Should C# be "better integrated" than, say, Rust or C++ via GDExtension.
I do think so, because C# is very suitable for a very huge selection of game developers, more so than native languages.
It is high level enough for rapid iterations & concise, readable code yet fast enough for very demanding simulations. The .NET runtime only keeps getting better and faster, so I would suggest that 95% of devs do not need more performance than what is possible with C# and Godot 4.X, going forward.

So, basically GDScript is the official, easy-to-use beginner-friendly integrated language, and C# the more advanced language for users who want to make more complex games that take several years to build.

Also, I personally think that (judging from that last survey Juan did), >20% devs using C# is actually a really large segment, given that the Godot Mono version is not marketed as the default, and having GDScript as the much more accessible first language.

I really wish for Godot to further embrace C# as a first-class language and keep it tightly integrated, as I think it fits absolutely perfectly and I would like to preferably work with this for the rest of my career as a game developer.

**Comment by neikeq** [comment]
> - Will .NET 6 be implemented in the same fashion as the Mono Runtime was implemented until now?

It has diverged completely from how we were doing things when embedding Mono. However, unlike what's described in this proposal, Godot is still the entry point of the application, which uses the .NET hosting APIs. I still plan to make it possible for the C# application to be the entry point. That will happen in Godot 4.1 or later, depending on its benefits (it may actually be a requirement if some .NET tools don't work well with custom .NET host apps).

> - Will this use the new GDExtension API, or will it continue to be a separate Editor binary like it was/is with Godot Mono?

Those two things are not related. There will only be one editor binary, which will include C# support (some dependencies may not be included by default, but the editor will download them transparently). No more Classic vs C# versions of Godot.

Regarding the GDExtension system, it's not clear yet whether C# will switch to it. My initial expectation was that GDExtension would introduce usability regressions compared to the scripting system, but the benefits would make it worth. However, later I found some serious issues that make it outright unsuitable for scripting (see godotengine/godot#57427, for example). Those issues may end up being fixed in time, but for now I'm playing safe and staying with the scripting system.

> - Is the plan to implement this for 4.0 or 4.1? (In my opinion waiting for 4.1 would be really disappointing for the substantial amount of devs currently using C#, ~20% is a lot and I reckon many of those would like to switch to 4.X as soon as possible)

This proposal is outdated, but mostly everything here is confirmed for Godot 4.0, except for the following:

- C# app as the entry point. As explained above, this will happen in Godot 4.1 or later.
- Mobile support. No decision made yet, I'm waiting for the maui release.

**Comment by paulloz** [comment]
> There will only be one editor binary, which will include C# support (some dependencies may not be included by default, but the editor will download them transparently). No more Classic vs C# versions of Godot.

This is IMHO a fantastic decision.

**Comment by Delpire** [comment]
@neikeq Will the scripting approach have an affect on debugging or performance monitoring compared to the GDExtension approach. My biggest pain points with Mono are that debugging requires special set up with either VSCode or Visual Studio (extensions, mono runtime, etc.), and performance monitoring with Mono seems to be broken on Windows. Another big pain is that inspecting variables while debugging sometimes doesn't work. My hope was that with .NET 6, these issues would just work natively with Visual Studio, likely simply through attaching to the process after running (I suppose having a C# app as the entry point in the future would make this even quicker).

**Comment by kopkako** [comment]
@neikeq Thanks so much for the writeup, I really appreciate it! ❤️
This makes me really happy and hopeful for the future of C# and Godot, I *love* that there will only be one editor going forward. This will also hopefully greatly unify the user base and benefit everyone.

Still trying to wrap my head around what it would mean for my dev workflow if the .NET app becomes the entry point instead of Godot. Would this have very heavy consequences for the way I would work with Godot, or is it more of a technical backend detail at the end of the day?

I am asking this because I am planning to build a rather involved Godot Mono tooling project, starting in 3.X and was hoping to eventually port that to 4.X – would that still be theoretically possible, or would everything break?

**Comment by jointri** [comment]
I'm a bit confused about something. If there will be only one binary release for 4.x and around 4.1 or so godot will have completed .NET 6 integration, does that mean I can use the engine's debugging features like I can with gdscript?

**Comment by GeorgeS2019** [comment]
@Calinou  Could you elaborate on why it is an incompatibility break?

**Comment by Calinou** [comment]
> @Calinou Could you elaborate on why it is an incompatibility break?

Switching from Mono to .NET Core will break things by design, since the APIs exposed to users will change to some extent. There's nothing that can be done to prevent this.

**Comment by GeorgeS2019** [comment]
In addition to the incompatibility break, I am curious about the information needed to complete the following table.

| Godot4 Support |  Branch |Android | iOS | Web | OpenXR | GDExtension| 
|----------------|---------|---------|-----|-----|--------|--------|
| **Mono**           |    Master    |         |     |     |        |        |
| **.NET6**          |  31.01.2022      |         |     |     |        |        |
| **.NET7**          |         |         |    |     |        |        |

### Q1: Is dotnet6 merge replacing mono a good idea?

> Currently it is possible to build Godot4 mono based on the latest master branch. It is unclear to me if the godot4 mono build supports desktop (windows and linux without issue), android, iOS, Web exports. Has anyone tested usability for Godot4 OpenXR.

> dotnet6 branch dependency on NativeAOT is now obsolete and has been replaced by .NET7. It is still a console app with an unclear timeline of when the NativeAOT support will come to mobile and web. 

> It is unclear how many users are currently actively developing using Godot4 mono. Do we need a survey or poll to decide if it is a good idea to replace mono with dotnet6 branch. Is it better to keep both existing in parallel? Is it possible to support building mono and dotnet? 

### Q2: what to do with existing Godot4 mono issues and PRs?
> Do we need discussion how to transition from mono to dotnet6? Currently, there are [53 issues](https://github.com/godotengine/godot/issues?q=is:open+is:issue+label:topic:dotnet+milestone:4.0) on Godot4 mono AND [14 PRs](https://github.com/godotengine/godot/pulls?q=is:pr+is:open++label:topic:dotnet+milestone:4.0+) on Godot4 mono. What does the community want to do with these issues and PRs. 

I do not have the overview and I hope the community steps forwards to discuss and share their views.

This topic is also being discussed [at Godot contributor forum](https://chat.godotengine.org/channel/scripting)

**Comment by GeorgeS2019** [comment]
There is an interesting [new Godot initiative ](https://github.com/godotengine/godot-proposals/issues/4773#issuecomment-1172838386)which could complement the dotnet6 branch

**Comment by chismar** [comment]
Ahoy, friends, long time no see. Can anyone do a little status update on how dotne6 integration into the Godot 4.0 going on? 
Any new ETAs?

**Comment by GeorgeS2019** [comment]
FYI: relevant to follow the latest feedback from @neikeq at Godot Contributor chat.

Source: [C# Godot 4.0 Roadmap](https://github.com/orgs/godotengine/projects/30/views/1?filterQuery=status%3A&visibleFields=%5B%22Title%22%2C%22Assignees%22%2C%22Status%22%2C%22Labels%22%5D)

![image](https://user-images.githubusercontent.com/49812372/182168173-6be01dc7-10ca-4310-89c7-bbf099427624.png)

![image](https://user-images.githubusercontent.com/49812372/182168572-85fbdd76-b3d8-4cd9-85bb-55769b338fd5.png)

**Comment by GeorgeS2019** [comment]
@Calinou 

Do you think it is appropriate to change the milestone from 4.1 to 4.0?

**Comment by GeorgeS2019** [comment]
### Other benefits:

#### I now have a better idea that this is related to **Editor Unification**
- [ ] Better IDE support by default, less features need to be implemented in Godot IDE extensions. Projects would be runnable by default as they would be executable rather than library C# projects. A Godot extension would only need to add the more fancy features like Play in Editor.

**Comment by KiwwiPity** [comment]
> # Motivation
> ### Other benefits:
> #### I now have a better idea that this is related to **Editor Unification**
> 
>     * [ ]  Better IDE support by default, less features need to be implemented in Godot IDE extensions. Projects would be runnable by default as they would be executable rather than library C# projects. A Godot extension would only need to add the more fancy features like Play in Editor.

Someone is trying to integrate (embedding) **VS Code** within Godot Editor:
https://www.reddit.com/r/godot/comments/wl3u86/embedding_vscode_directly_into_the_godot_editor/

**Comment by GeorgeS2019** [comment]
@KiwwiPity   This means, beside GDScript, one can use the embeded VSCode for e.g. python, rust, go etc. in addition to C# and other .NET langauges

**Comment by KiwwiPity** [comment]
Yes I know. But isn't it interesting for C#?
I thought that integrate C# on the Script Workspace were in the roadmap, now that many Unity users are migrating to us.

**Comment by paulloz** [comment]
> now that many Unity users are migrating to us

This really doesn't change anything. Anybody coming from Unity is already used to having an IDE for code on the one side and the engine on the other, since the engine (imho rightfully) doesn't provide an integrated one.

**Comment by RedMser** [comment]
@KiwwiPity Do note that this is a very hacky approach which is never intended to be an "officially endorsed" solution. I wouldn't point newcomers to the engine to this extension, since it breaks some core functionality (like debugging from within Godot, double-click to connect signals, shortcuts are captured by the IDE, can't view documentation from editor, etc.).

I'd rather see Godot's own code editor improved further as a long-term solution (such as with hover hints https://github.com/godotengine/godot/pull/63908 , better autocomplete search, splitting windows etc.).

**Comment by KiwwiPity** [comment]
> > now that many Unity users are migrating to us
> 
> This really doesn't change anything. Anybody coming from Unity is already used to having an IDE for code on the one side and the engine on the other, since the engine (imho rightfully) doesn't provide an integrated one.

Didn't knew that Unity hasn't an IDE

**Comment by Shadowblitz16** [comment]
> > > now that many Unity users are migrating to us
> > 
> > 
> > This really doesn't change anything. Anybody coming from Unity is already used to having an IDE for code on the one side and the engine on the other, since the engine (imho rightfully) doesn't provide an integrated one.
> 
> Didn't knew that Unity hasn't an IDE

it doesn't have a official build in editor ide but you can download one for free or buy one

**Comment by GeorgeS2019** [comment]
# What could be NEXT after dotnet6 merged?

The proposal spoke about improving Visual Studio IDE experience. Here we are making debugging using Visual Studio (e.g. 2022) more reliable. Please bring WHAT YOU THINK should be improved in e.g. VS2022 experience to this [discussion](https://github.com/godotengine/godot-csharp-visualstudio/issues/36#issuecomment-1233641112).

**Comment by daerogami** [comment]
That's really it.

I just want to be able to start the project from my own entry class so I can do other cool stuff like using the ecosystem around `Microsoft.Extensions.Hosting.IHostBuilder`. Is it the right tool for the job? Probably not, but if the IDE experience was that seamless, that would be wicked neat to call something like `builder.AddGodotEngine()` and have DI ready to go out-of-the-box.

**Comment by neikeq** [comment]
Regarding the C# application being the entry-point, it's still something I would like to see, but it's definitely not happening in 4.0 and likely not in 4.1 either, which will focus on mobile (although the mobile port may work that way).

Most of the other things mentioned here are already done or WIP, and they are coming in Godot 4.0. I'm going to close this proposal as it's old and covered a lot of things. Further discussion would be better tracked in separate threads.

Before it's inclusion in an official release, I plan to make a blog post detailing what's new for C# in Godot 4.0.

**Comment by GeorgeS2019** [comment]
> they are coming in Godot 4.0

**Comment by KiwwiPity** [comment]
@GeorgeS2019 hi there
what's going on with Godot 4 C# progress, there's some thread to keep track of the progress and so?

**Comment by tvardero** [comment]
I don't think this issue should have been closed back then. As we haven't _yet_ achieved Godot as a library.

Also comes to think:
* Silk.Net
   * Silk.Net.Math for generic vectors;
   * Silk.Net.Input for window inputs and mouse cursors;
   * Silk.Net.Windowing for windows management;
* Microsoft.Extensions.Hosting, which comes with:
   * (shortened) MS.Ext.Logging
   * MS.Ext.Configuration
   * MS.Ext.DI
   * MS.Ext.Options
 
All that is Ms.Ext.* - is half-abstraction, half-default implementation, you can use Autofac, NInject, Serilog, NLog and whatever you want on top of Ms.Ext.*

**Comment by GeorgeS2019** [comment]
@tvardero 

Many of us move on to this [c# discussion](https://github.com/godotengine/godot-proposals/discussions/4007)

You are right with Silk.Net

Godot4 team is working HARD on DirectX driver. It is working now 90 to 95%. This needs more people to test. 

Once that is done,  Godot4 with .NET6 has a good chance to address one of the major gaps in .NET 3D ecosystem. 

The lack of mature MIT license 3D engine, non Unity, to meet WinUI and Maui and Blazor Business requirements.

Please open new issues and draw attentions through here and Godot Discords focusing on C#/.NET

**Comment by Calinou** [comment]
Using Godot as a C# library is likely dependent on https://github.com/godotengine/godot-proposals/issues/4773 being implemented first.

**Comment by GeorgeS2019** [comment]
@Calinou  thanx for pointing out that it is **4.x**

![image](https://user-images.githubusercontent.com/49812372/214808727-7036f068-4da2-44c1-b315-699068b02129.png)
